chainerx_cc/chainerx/routines/linalg.cc

```diff
@@ -189,4 +192,122 @@ Array Inverse(const Array& a) {
     return out;
 }
 
+std::tuple<Array, Array, Array> SVD(const Array& a, bool full_matrices, bool compute_uv) {
+    CheckRankTwoArray(a);
+    Array u{};
+    Array s{};
+    Array v{};
+
+    {
+        NoBackpropModeScope scope{};
+        std::tie(u, s, v) = a.device().backend().CallKernel<SVDKernel>(a, full_matrices, compute_uv);
+    }
+
+    // Reference:
+    // https://j-towns.github.io/papers/svd-derivative.pdf
+    {
+        BackwardBuilder bb{"svd", a, {u, s, v}};
+        if (BackwardBuilder::Target bt = bb.CreateTarget(0)) {
+            bt.Define([a_tok = bb.RetainInput(0),
+                       u_tok = bb.RetainOutput(0),
+                       s_tok = bb.RetainOutput(1),
+                       v_tok = bb.RetainOutput(2),
+                       full_matrices,
+                       compute_uv](BackwardContext& bctx) {
+                if (full_matrices) {
+                    throw ChainerxError{"ChainerX SVD differentiation is not implemented for full_matrices=true."};
+                }
+                if (!compute_uv) {
+                    throw ChainerxError{"ChainerX SVD differentiation cannot be computed with compute_uv=false."};
+                }
+
+                const Array& a = bctx.GetRetainedInput(a_tok);
+                const Array& u = bctx.GetRetainedOutput(u_tok);
+                const Array& s = bctx.GetRetainedOutput(s_tok);
+                const Array& vt = bctx.GetRetainedOutput(v_tok);
+
+                auto m = a.shape()[0];
+                auto n = a.shape()[1];
+                auto k = s.shape()[0];
+
+                const Array& gu = bctx.output_grad(0).has_value() ? *bctx.output_grad(0) : Zeros(u.shape(), a.dtype(), a.device());
+                const Array& gsigma = bctx.output_grad(1).has_value() ? *bctx.output_grad(1) : Zeros(s.shape(), a.dtype(), a.device());
+                const Array& gvt = bctx.output_grad(2).has_value() ? *bctx.output_grad(2) : Zeros(vt.shape(), a.dtype(), a.device());
+
+                auto v = vt.Transpose();
+                auto gv = gvt.Transpose();
+
+                auto sigma_term = Dot(Dot(u, Diag(gsigma)), vt);
+
+                auto ut = u.Transpose();
+                auto im = Eye(m, m, 0, a.dtype(), a.device());
+                auto in = Eye(n, n, 0, a.dtype(), a.device());
+                auto sigma_mat = Diag(s);
+                auto sigma_mat_inv = Diag(Power(s, -1));
+                auto sigma_sq = Power(s, 2);
+                auto F = ExpandDims(sigma_sq, 0) - ExpandDims(sigma_sq, 1);
+                // Invert values of F, and fill the diagonal with 0s.
+                // F has 0s on the diagonal, therefore fill it first with infinity.
+                auto mask = Eye(F.shape()[0], F.shape()[1], 0, Dtype::kBool, a.device());
```

Comment:
I personally find these `auto`s fine but it goes against our convention, could you be explicit with the types in this lambda?